p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,390 @@
+    name: { type: \"str\", value: tstr }
+) }";
+
+/// An iterator struct for Document.
+#[derive(Debug)]
+pub struct DocumentIter(Vec<OperationWithMeta>);
+
+/// A resolvable data type made up of a collection of causally linked operations.
+#[derive(Debug, Clone)]
+pub struct Document {
+    /// The create operation which is the root of this document.
+    create_operation: OperationWithMeta,
+    /// A causal graph of this documents operations which can be topologically sorted.
+    graph: Graph<OperationWithMeta>,
+}
+
+impl Document {
+    /// The hash id of this document.
+    pub fn id(&self) -> Hash {
+        self.create_operation.operation_id().to_owned()
+    }
+
+    /// The hash id of this documents schema.
+    pub fn schema_hash(&self) -> Hash {
+        self.create_operation.schema()
+    }
+
+    /// The author of this document.
+    pub fn author(&self) -> Author {
+        self.create_operation.public_key().to_owned()
+    }
+
+    /// The schema for this document.
+    pub fn schema(&self) -> Schema {
+        Schema::new(&self.schema_hash(), DOCUMENT_SCHEMA).unwrap()
+    }
+
+    /// Returns an iterator over all operations in this document ordered topologically.
+    pub fn iter(&self) -> Result<DocumentIter, DocumentError> {
+        let sorted = match self.graph.sort() {
+            Ok(operations) => Ok(operations.into_iter().rev().collect()),
+            Err(e) => Err(DocumentError::GraphSortingError(e)),
+        }?;
+
+        Ok(DocumentIter(sorted))
+    }
+
+    /// Sort the graph topologically, then reduce the linearised operations into a single
+    /// `Instance`.
+    pub fn resolve(&self) -> Result<Instance, DocumentError> {
+        let mut document_iter = self.iter()?;
+
+        let create_message = document_iter.next().unwrap();
+        let mut instance = Instance::try_from(create_message)?;
+
+        document_iter.try_for_each(|op| instance.apply_update(op))?;
+
+        Ok(instance)
+    }
+}
+
+impl Iterator for DocumentIter {
+    type Item = OperationWithMeta;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.pop()
+    }
+}
+
+/// A struct for building documents.
+#[derive(Debug)]
+pub struct DocumentBuilder {
+    /// An unsorted collection of operations which are associated with a particular document id.
+    operations: Vec<OperationWithMeta>,
+}
+
+impl DocumentBuilder {
+    /// Instantiate a new DocumentBuilder with a collection of operations.
+    pub fn new(operations: Vec<OperationWithMeta>) -> Self {
+        Self { operations }
+    }
+
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
```

Comment:
See comment above! We could pass in the `schema_definition` here (which would then be validated against the given `schema_hash` as @cafca noted.

Reply:
Yes, thank you both! This is much better :+1:

Reply:
Woho! Then we can finally make `DOCUMENT_SCHEMA` an actual test value :partying_face: